compaction.go

```diff
@@ -0,0 +1,195 @@
+package frostdb
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log/level"
+	"github.com/parquet-go/parquet-go"
+
+	"github.com/polarsignals/frostdb/dynparquet"
+	"github.com/polarsignals/frostdb/parts"
+)
+
+var ErrCompactionRecoveryFailed = fmt.Errorf("failed to recover compacted level")
+
+type FileCompaction struct {
+	t      *Table
+	file   *os.File
+	offset int64 // Writing offsets into the file
+	ref    int64 // Number of references to file.
+}
+
+func (f *FileCompaction) Close() error {
+	return f.file.Close()
+}
+
+func NewFileCompaction(t *Table, lvl int) *FileCompaction {
+	f := &FileCompaction{
+		t: t,
+	}
+
+	if t.db.storagePath == "" { // No storage path provided, use temp files.
+		file, err := os.CreateTemp("", fmt.Sprintf("L%v-*.parquet", lvl))
+		if err != nil {
+			panic(err)
+		}
+		level.Info(t.logger).Log("msg", "created temp file for level", "level", lvl, "file", file.Name())
+		f.file = file
+		return f
+	}
+
+	indexDir := filepath.Join(t.db.indexDir(), t.name)
+	if err := os.MkdirAll(indexDir, dirPerms); err != nil {
+		panic(err)
+	}
+
+	file, err := os.OpenFile(filepath.Join(indexDir, fmt.Sprintf("L%v.parquet", lvl)), os.O_CREATE|os.O_RDWR, filePerms)
+	if err != nil {
+		panic(err)
+	}
+	f.file = file
+	return f
+}
+
+// accountingWriter is a writer that accounts for the number of bytes written.
+type accountingWriter struct {
+	w io.Writer
+	n int64
+}
+
+func (a *accountingWriter) Write(p []byte) (int, error) {
+	n, err := a.w.Write(p)
+	a.n += int64(n)
+	return n, err
+}
+
+// writeRecordsToParquetFile will compact the given parts into a Parquet file written to the next level file.
+func (f *FileCompaction) writeRecordsToParquetFile(compact []parts.Part, options ...parts.Option) ([]parts.Part, int64, int64, error) {
+	if len(compact) == 0 {
+		return nil, 0, 0, fmt.Errorf("no parts to compact")
+	}
+
+	accountant := &accountingWriter{w: f.file}
+	preCompactionSize, err := f.t.compactParts(accountant, compact,
+		parquet.KeyValueMetadata(
+			"compaction_tx", // Compacting up through this transaction.
+			fmt.Sprintf("%v", compact[0].TX()),
+		),
+	) // compact into the next level
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	// Record the writing offset into the file.
+	prevOffset := f.offset
+	f.offset += accountant.n + 8
+	f.ref++
+
+	// Record the file size for recovery.
+	size := make([]byte, 8)
+	binary.LittleEndian.PutUint64(size, uint64(accountant.n))
+	if n, err := f.file.Write(size); n != 8 {
```

Comment:
So I've been thinking about this and with the failure recovery that we have we fallback to recovering from the wal. However, in the case where we truncate the wal after snapshot I think it would make sense to fsync the index files.

Reply:
Added an fsync step before wal truncation